stellargraph/layer/knowledge_graph.py

```diff
@@ -529,6 +529,164 @@ def in_out_tensors(self):
+        rel += np.cos(phase)
+
+        return node, rel
+
+    def rank_edges_against_all_nodes(
+        self, test_data, known_edges_graph, tie_breaking="random"
+    ):
+        if not isinstance(test_data, KGTripleSequence):
+            raise TypeError(
+                "test_data: expected KGTripleSequence; found {type(test_data).__name__}"
+            )
+
+        num_nodes = known_edges_graph.number_of_nodes()
+
+        all_node_embs, all_rel_embs = self.embeddings()
+
+        raws = []
+        filtereds = []
+
+        # run through the batches and compute the ranks for each one
+        num_tested = 0
+        for ((subjects, rels, objects),) in test_data:
+            num_tested += len(subjects)
+
+            # batch_size x k
+            ss = all_node_embs[subjects, :]
+            rs = all_rel_embs[rels, :]
+            os = all_node_embs[objects, :]
+
+            # (the margin is a fixed offset that doesn't affect relative ranks)
+            mod_o_pred = -np.linalg.norm(
+                (ss * rs)[None, :, :] - all_node_embs[:, None, :], axis=2
+            )
+            mod_s_pred = -np.linalg.norm(
+                (all_node_embs)[:, None, :] * rs[None, :, :] - os[None, :, :], axis=2
+            )
+
+            mod_o_raw, mod_o_filt = _ranks_from_score_columns(
+                mod_o_pred,
+                true_modified_node_ilocs=objects,
+                unmodified_node_ilocs=subjects,
+                true_rel_ilocs=rels,
+                modified_object=True,
+                known_edges_graph=known_edges_graph,
+                tie_breaking=tie_breaking,
+            )
+            mod_s_raw, mod_s_filt = _ranks_from_score_columns(
+                mod_s_pred,
+                true_modified_node_ilocs=subjects,
+                true_rel_ilocs=rels,
+                modified_object=False,
+                unmodified_node_ilocs=objects,
+                known_edges_graph=known_edges_graph,
+                tie_breaking=tie_breaking,
+            )
+
+            raws.append(np.column_stack((mod_o_raw, mod_s_raw)))
+            filtereds.append(np.column_stack((mod_o_filt, mod_s_filt)))
+
+        # make one big array
+        raw = np.concatenate(raws)
+        filtered = np.concatenate(filtereds)
+        # for each edge, there should be an pair of raw ranks
+        assert raw.shape == filtered.shape == (num_tested, 2)
+
+        return raw, filtered
+
+    def __call__(self, x):
+        s_iloc, r_iloc, o_iloc = x
+
+        s_re = self._node_embeddings_real(s_iloc)
+        s_im = self._node_embeddings_imag(s_iloc)
+
+        r_phase = self._edge_type_embeddings_phase(r_iloc)
+
+        r_re = tf.math.cos(r_phase)
+        r_im = tf.math.sin(r_phase)
+
+        o_re = self._node_embeddings_real(o_iloc)
+        o_im = self._node_embeddings_imag(o_iloc)
+
+        return self._scoring([s_re, s_im, r_re, r_im, o_re, o_im])
+
+    def in_out_tensors(self):
+        s_iloc = Input(shape=1)
```

Comment:
Is this extra dimension `shape=1` needed here? From what I can tell this dimension isn't being used, could this instead be:

```suggestion
        s_iloc = Input(batch_shape=(None,))
``

Reply:
That's potentially true, although it makes this differ to `ComplEx` and `DistMult`. As such, I'd prefer to keep it as is, and change them all separately?

Reply:
yeah fair point, just something to flag if you're gonna go ahead with that refactor of all 3 but its pretty minor :)

Reply:
I made a note on #1549.